crates/noirc_evaluator/src/ssa_refactor/opt/defunctionalize.rs

```diff
@@ -173,72 +159,129 @@ impl DefunctionalizationContext {
     }
 
     /// Returns the apply function for the given signature
-    fn get_apply_function(&self, signature: &FunctionSignature) -> ApplyFunction {
+    fn get_apply_function(&self, signature: &Signature) -> ApplyFunction {
         *self.apply_functions.get(signature).expect("Could not find apply function")
     }
 }
 
-/// Collects all functions used as a value by their signatures
-fn find_variants(ssa: &Ssa) -> HashMap<FunctionSignature, Vec<FunctionId>> {
-    let mut variants: HashMap<FunctionSignature, Vec<FunctionId>> = HashMap::new();
-    let mut functions_used_as_values = HashSet::new();
+/// Collects all functions used as values that can be called by their signatures
+fn find_variants(ssa: &Ssa) -> HashMap<Signature, Vec<FunctionId>> {
+    let mut dynamic_dispatches: HashSet<Signature> = HashSet::new();
+    let mut functions_as_values: HashSet<FunctionId> = HashSet::new();
 
```

Comment:
Wonder if this code could be more abstract, e.g. an abstraction over HashMap that has builder methods used here, but good with this for this pass